# Multiple Linear Regression

In [1]:
import numpy as np
import pandas as pd

In [2]:
from sklearn.datasets import load_diabetes

In [3]:
X, y = load_diabetes(return_X_y = True)

In [4]:
X.shape

(442, 10)

In [5]:
y.shape

(442,)

Thus our dataset has 10 predictors and 1 input column. This dataset has all the features as numerical columns and thus this is a perfect candidate for
multiple linear regression. Since we have 10 input features we need to find 11 values $\beta_0, \beta_1, \ldots, \beta_{10}$
 in order to determine the equation of our hyperplane.

$$
\beta = \begin{pmatrix}
\beta_0 \\
\beta_1 \\
\beta_2 \\
\vdots \\
\beta_{10}
\end{pmatrix}, \quad \beta \in \mathbb{R}^{11}
$$


### Let's take a look at the dataset

In [6]:
X

array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
         0.01990749, -0.01764613],
       [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
        -0.06833155, -0.09220405],
       [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
         0.00286131, -0.02593034],
       ...,
       [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
        -0.04688253,  0.01549073],
       [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
         0.04452873, -0.02593034],
       [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
        -0.00422151,  0.00306441]])

In [7]:
y

array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
        69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
        68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
        87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
       259.,  53., 190., 142.,  75., 142., 155., 225.,  59., 104., 182.,
       128.,  52.,  37., 170., 170.,  61., 144.,  52., 128.,  71., 163.,
       150.,  97., 160., 178.,  48., 270., 202., 111.,  85.,  42., 170.,
       200., 252., 113., 143.,  51.,  52., 210.,  65., 141.,  55., 134.,
        42., 111.,  98., 164.,  48.,  96.,  90., 162., 150., 279.,  92.,
        83., 128., 102., 302., 198.,  95.,  53., 134., 144., 232.,  81.,
       104.,  59., 246., 297., 258., 229., 275., 281., 179., 200., 200.,
       173., 180.,  84., 121., 161.,  99., 109., 115., 268., 274., 158.,
       107.,  83., 103., 272.,  85., 280., 336., 281., 118., 317., 235.,
        60., 174., 259., 178., 128.,  96., 126., 28

## Using the scikit-learn multiple linear regression model

In [8]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split as tts

In [9]:
sklearn_model = LinearRegression()
X_train, X_test, y_train, y_test = tts(X, y, test_size = 0.3, random_state = 40)

In [10]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((309, 10), (133, 10), (309,), (133,))

In [11]:
sklearn_model.fit(X_train, y_train)

LinearRegression()

In [12]:
sklearn_model.coef_

array([  15.54992881, -229.93487166,  582.20927562,  282.72195707,
       -729.21739561,  359.41016942,   27.98870671,  160.23780607,
        768.50303028,   32.77695593])

Thus

$$
\beta = \begin{pmatrix}
\beta_1 \\
\beta_2 \\
\beta_3 \\
\beta_4 \\
\beta_5 \\
\beta_6 \\
\beta_7 \\
\beta_8 \\
\beta_9 \\
\beta_{10}
\end{pmatrix}
=
\begin{pmatrix}
15.54992881 \\
-229.93487166 \\
582.20927562 \\
282.72195707 \\
-729.21739561 \\
359.41016942 \\
27.98870671 \\
160.23780607 \\
768.50303028 \\
32.77695593
\end{pmatrix}, \quad \beta \in \mathbb{R}^{10}
$$


In [13]:
sklearn_model.intercept_

150.1026884474012

Thus, 
$$
\beta_0 = 150.1026884474012
$$

and hence we have found our $\beta$ vector as follows:

$$
\beta = \begin{pmatrix}
\beta_0 \\
\beta_1 \\
\beta_2 \\
\beta_3 \\
\beta_4 \\
\beta_5 \\
\beta_6 \\
\beta_7 \\
\beta_8 \\
\beta_9 \\
\beta_{10}
\end{pmatrix}
=
\begin{pmatrix}
150.1026884474012 \\
15.54992881 \\
-229.93487166 \\
582.20927562 \\
282.72195707 \\
-729.21739561 \\
359.41016942 \\
27.98870671 \\
160.23780607 \\
768.50303028 \\
32.77695593
\end{pmatrix}, \quad \beta \in \mathbb{R}^{11}
$$


## Checking the performance of sklearn_model

In [14]:
from sklearn.metrics import r2_score

In [15]:
r2_score(y_test, sklearn_model.predict(X_test))

0.40381960432354824

# Creating our own custom linear regression model class

In [16]:
from numpy.linalg import LinAlgError
class MeraMLR:
    def __init__(self):
        self.coef_ = None
        self.intercept_ = None
        self.beta = None

    def fit(self, X_train, y_train):
        X = np.insert(X_train, obj = 0, values = 1, axis = 1) # add a column with all the values as 1 inside the X_train
        y = y_train
        
        try:
            β = np.linalg.inv(X.T @ X) @ X.T @ y
        except LinAlgError as ex:
            print(ex)

        self.coef_ = β[1::]
        self.intercept_ = β[0]
        self.beta = β

    def predict(self, X_test):
        X = np.insert(X_test, obj = 0, values = 1, axis = 1) # add a column with all the values as 1 inside the X_train
        return X @ self.beta

In [17]:
my_MLR_model = MeraMLR()

In [18]:
my_MLR_model.fit(X_train, y_train)

# Comparing our MLR class with that of sklearn class

In [19]:
epsilon = 0.0000001

sklearn_model_coefficients = sklearn_model.coef_
my_model_coefficients = my_MLR_model.coef_

In [20]:
flag = True
for x, y in zip(sklearn_model_coefficients, my_model_coefficients):
    if x-y >= epsilon:
        flag = False
        break

if flag:
    print('Our model coefficients == sklearn model coefficients')
else:
    print('Our model coefficients != sklearn model coefficients😢')

Our model coefficients == sklearn model coefficients


Similiarly, we can check the intercepts also.

In [21]:
if abs(sklearn_model.intercept_ - my_MLR_model.intercept_) < epsilon:
    print('our model intercept == sklearn model intercept')
else:
    print('our model intercept != sklearn model intercept 😢')

our model intercept == sklearn model intercept


# Comparing the r2 scores

In [22]:
r2_score_sklearn_model = r2_score(sklearn_model.predict(X_test), y_test)

In [23]:
r2_score_my_model = r2_score(my_MLR_model.predict(X_test), y_test)

In [24]:
r2_score_sklearn_model, r2_score_my_model

(-0.01574474947378146, -0.015744749473783237)

# Comparing the predictions of both the models

In [25]:
my_model_predictions = my_MLR_model.predict(X_test)
sklearn_model_predictions = sklearn_model.predict(X_test)

In [26]:
for x, y in zip(my_model_predictions, sklearn_model_predictions):
    if abs(x-y) >= epsilon:
        print("my model predictions != sklearn model predictions 😢😢")
        break
else:
    print("my model predictions ==  sklearn model predictions 🎉🎉")

my model predictions ==  sklearn model predictions 🎉🎉
